staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	}
+
+	// No labels or NodeAffinity exist, default to GA version
+	return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+}
+
+// TopologyKeyExist checks if a certain key exists in a VolumeNodeAffinity
+func TopologyKeyExist(key string, vna *v1.VolumeNodeAffinity) bool {
+	if vna == nil || vna.Required == nil || vna.Required.NodeSelectorTerms == nil || len(vna.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	for _, nodeSelectorTerms := range vna.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+		for _, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+type regionTopologyHandler func(pv *v1.PersistentVolume) error
+
+// translateTopologyFromCSIToInTree translate a CSI topology to
+// Kubernetes topology and add topology labels to it. Note that this function
+// will only work for plugin with single topologyKey. If a plugin has
+// more than one topologyKey, it will need to be processed separately
+// by the plugin.
+// regionTopologyHandler is a function to generate region topology NodeAffinity
+// for the given PV. It assumes the Zone NodeAffinity already exists
+// If regionTopologyHandler is nil, no region NodeAffinity will be added
+// If regionTopologyHandler returns err, it will be ignored.
+//
+// 1. Replace all CSI topology to Kubernetes Zone topology label, this including
+//    combine existing kubernetes zone together
+// 2. Process and generate region topology if a regionTopologyHandler is passed
+// 3. Add Kubernetes Topology labels(zone and region) if they do not exist
+func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey string, regionTopologyHandler regionTopologyHandler) error {
+
+	_, zoneLabel, regionLabel := getTopologyLabel(pv)
+
+	// 1. Replace all CSI topology to Kubernetes Zone label
+	err := replaceTopology(pv, topologyKey, zoneLabel)
+	if err != nil {
+		return fmt.Errorf("Failed to replace CSI topology to Kubernetes topology, error: %v", err)
+	}
+
+	// 2. Take care of region topology if a regionTopologyHandler is passed
+	if regionTopologyHandler != nil {
+		// let's make less strict on this one. Even if there is an error in the region processing, just ignore it
+		_ = regionTopologyHandler(pv)
+	}
+
+	// 3. Add labels about Kubernetes Topology
+	zoneVals := getTopologyValues(pv, zoneLabel)
+	if len(zoneVals) > 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		_, zoneOK := pv.Labels[zoneLabel]
+		if !zoneOK {
+			zoneValStr := strings.Join(zoneVals, labelMultiZoneDelimiter)
+			pv.Labels[zoneLabel] = zoneValStr
+		}
+	}
+
+	regionVals := getTopologyValues(pv, regionLabel)
+	if len(regionVals) > 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		_, regionOK := pv.Labels[regionLabel]
+		if !regionOK {
+			regionValStr := strings.Join(regionVals, labelMultiZoneDelimiter)
```

Comment:
Good question. This is just in case some crazy driver provides ability like multiple terms:
1. region A, zone 1
2. region B, zone 2

If it only return one regionVals it will do no harm to GCP so I think it should be fine.

Reply:
I think adding region labels should be the responsibility of the region handler. I'm not sure if any volume type we're dealing with supports multi region.

Reply:
Okay, I will move this logic to gce then.